File: src/amtc_utils/NeroResourceClient.cpp
//
// Created by finostro on 27-11-23.
//
#include <amtc_utils/nero_resource/NeroResourceClient.h>



NeroResourceClient::NeroResourceClient(rclcpp::Node *node, const char* resource_type)
{
  node_ = node;
  resource_type_ = resource_type;
  access_token_ = "";
  callback_group_ = node_->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
  alloc_client_ = node_->create_client<resource_manager_msgs::srv::Alloc>("resource_manager/alloc", rmw_qos_profile_services_default, callback_group_);
  free_client_ = node_->create_client<resource_manager_msgs::srv::Free>("resource_manager/free", rmw_qos_profile_services_default, callback_group_);

}

bool NeroResourceClient::alloc(const rclcpp::Duration &timeout) {

    auto request = std::make_shared<resource_manager_msgs::srv::Alloc::Request>();
    request->resource.type = resource_type_;
    auto future = alloc_client_->async_send_request(request);
    auto future_status = future.wait_for(timeout.to_chrono<std::chrono::duration<int64_t, std::milli> >());

    if (future_status == std::future_status::timeout)
    {
        RCLCPP_ERROR(node_->get_logger(), "Failed to call service resource_manager/alloc");

        return false;
    }
    else{
      auto result = future.get();
      if(!result->is_success)
        {
            RCLCPP_ERROR(node_->get_logger(), "Failed to allocate resource");
            return false;
        }
        access_token_ = result->token.data;
        return true;

    }
}

bool NeroResourceClient::free(const rclcpp::Duration &timeout) {
    auto request = std::make_shared<resource_manager_msgs::srv::Free::Request>();
    request->token.data = access_token_;
    auto result = free_client_->async_send_request(request);
    auto future_status = result.wait_for(timeout.to_chrono<std::chrono::duration<int64_t, std::milli> >());

    if (future_status == std::future_status::ready)
    {
        if(!result.get()->is_success)
        {
            RCLCPP_ERROR(node_->get_logger(), "Failed to free resource");
            return false;
        }

        access_token_ = "";

        return true;
    }
    else
    {
        RCLCPP_ERROR(node_->get_logger(), "Failed to call service resource_manager/free");
        return false;
    }
}
bool NeroResourceClient::is_allocated() {
    return !access_token_.empty();
}

File: amtc_utils/nero_resource_client.py
#!/usr/bin/env python3

from rclpy.node import Node
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup

from resource_manager_msgs.srv._alloc import Alloc
from resource_manager_msgs.srv._free import Free
from resource_manager_msgs.msg import Token


class NeroResourceClient:

    def __init__(
            self,
            node: Node,
            resource_type: str,
            callback_group: MutuallyExclusiveCallbackGroup
    ) -> None:
        self.node = node
        self.resource_type = resource_type
        self.access_data = Token()
        self._srv_call_cbg = callback_group

        # Namespace
        self._namespace = node.get_namespace()

        # Service clients
        self._cli_alloc_resource = node.create_client(srv_type=Alloc,
                                                      srv_name='resource_manager/alloc',
                                                      callback_group=self._srv_call_cbg)

        self._cli_free_resource = node.create_client(srv_type=Free,
                                                     srv_name='resource_manager/free',
                                                     callback_group=self._srv_call_cbg)

    def alloc(
            self,
            timeout: float
    ) -> bool:
        """
        Calls resource manager to allocate resource named during object creation.

        :param timeout: Time to wait for allocation service.
        :type timeout: float
        :return: True if resource successfully allocated, False else.
        :rtype: bool
        """

        alloc_req = Alloc.Request()
        alloc_req.resource.type = self.resource_type

        if not self._cli_alloc_resource.wait_for_service(timeout_sec=timeout):
            self.node.get_logger().info('[RESOURCE {0}] Failed because alloc server timeout...')
            return False

        future = self._cli_alloc_resource.call_async(alloc_req)
        self.node.executor.spin_until_future_complete(future)

        if future.result().is_success is False:
            self.node.get_parameter('[RESOURCE {0}] Failed to allocate resource {0}'.format(self.resource_type))
            return False

        self.access_data.data = future.result().token.data
        return True

    def free(
            self,
            timeout: float
    ) -> bool:
        """
        Calls resource manager to free resource named during object creation.

        :param timeout: Time to wait for freeing service.
        :type timeout: float
        :return: True if resource successfully freed, False else.
        :rtype: bool
        """

        free_req = Free.Request()
        free_req.token.data = self.access_data.data

        if not self._cli_free_resource.wait_for_service(timeout_sec=timeout):
            self.node.get_logger().info('[RESOURCE {0}] Failed because free server timeout...')
            return False

        future = self._cli_free_resource.call_async(free_req)
        self.node.executor.spin_until_future_complete(future)

        if future.result().is_success is False:
            self.node.get_parameter('[RESOURCE {0}] Failed to free resource {0}'.format(self.resource_type))
            return False

        self.access_data.data = ""
        return True
